lib/crypto-primitives/src/Cryptography/KDF/PBKDF2.hs

```diff
@@ -1,13 +1,140 @@
+-- | key=F4CA507C07D0BD31BC779A08756826A6FD9DD97D43AC25E4B29A0933ABEA03F3
+-- | -> fst $ generateKey SHA512 10000 (32,0) "password" Nothing
+-- |
+-- | 4. 'key' and 'iv' needed for aes-256-cbc with sha256 and salt='00000000' (which is in hex equivalent to '3030303030303030')
+-- | $ echo -n "metadata-secret" | openssl enc -aes-256-cbc -pbkdf2 -pass stdin -P -S 3030303030303030 -iter 10000 -P
+-- | salt=3030303030303030
+-- | key=57FCB522B950BCB78138EECFE7FBE07881E5B49AAA2D1CD761D4495A09A5F16C
+-- | iv =41D88E094C8F202C7DF6654B7F40E5AF
+-- | generateKey SHA512 10000 (32,16) "metadata-secret" (Just "00000000")
+generateKey
+    :: HashAlgorithm h
+    => h
+    -- ^ hash algorithm type, like SHA128, SHA256, SHA512
+    -> Word16
+    -- ^ number of iterations algorithm runs through, more iterations mean slower
+    -> (Word8, Word8)
+    -- ^ key length in bytes, iv length in bytes
+    -> ByteString
+    -- ^ payload
+    -> Maybe ByteString
+    -- ^ salt
+    -> (ByteString, ByteString)
+    -- ^ (key, iv)
+generateKey hashAlgoType iterations (keyLen, ivLen) payload saltM =
+    BS.splitAt (fromIntegral keyLen) whole
+  where
+    whole = KDF.generate
+        (KDF.prfHMAC hashAlgoType)
+        (KDF.Parameters (fromIntegral iterations) (fromIntegral $ keyLen + ivLen))
+        payload
+        (fromMaybe BS.empty saltM)
+
+generateKeyByron
+    :: ByteString
+    -- ^ payload
+    -> ByteString
+generateKeyByron key =
+    fst $ generateKey SHA512 500 (32,0) key (Just salt)
+  where
+    salt = "address-hashing" :: ByteString
+
+generateKeyIcarus
+    :: ByteString
+    -- ^ payload
+    -> ByteString
+generateKeyIcarus key =
+    fst $ generateKey SHA512 2048 (64,0) key (Just salt)
+  where
+    salt = "mnemonic" :: ByteString
+
+-- | Key Derivation Function compliant with https://cips.cardano.org/cips/cip83/
+-- | According to the CIP there are the following parameters for KDF:
+-- |  |------------------------------------------|
+-- |  | cipher	     aes-256-cbc (salted)        |
+-- |  | digest	     pdkdf2                      |
+-- |  | iterations	 10000 (default)             |
+-- |  | key + iv	 32 bytes key + 16 bytes iv  |
+-- |  | salt	     8 bytes                     |
+-- |  |------------------------------------------|
+-- |
+-- | Example:
+-- | $ echo -n "metadata-secret" | openssl enc -aes-256-cbc -pbkdf2 -md sha256 -pass stdin -P -S 3030303030303030 -iter 10000 -P
+-- | salt=3030303030303030
+-- | key=57FCB522B950BCB78138EECFE7FBE07881E5B49AAA2D1CD761D4495A09A5F16C
+-- | iv =41D88E094C8F202C7DF6654B7F40E5AF
+-- |
+-- | which is equivalent to (openSSL requires hex decoded salt):
+-- | -> generateKeyMetadata "metadata-secret" (Just "00000000")
+generateKeyMetadata
+    :: ByteString
+    -- ^ payload
+    -> Maybe ByteString
+    -- ^ salt
+    -> (ByteString, ByteString)
+generateKeyMetadata =
+    generateKey SHA256 10000 (32,16)
```

Comment:
```suggestion
```

These numerical values should not be a concern of this particular module. See other comment.